include/graphblas/reference/blas3.hpp

```diff
@@ -1111,92 +1118,545 @@ namespace grb {
+					<< "exclusive\n";
+#endif
+				return ILLEGAL;
+			}
+
+			if( getID(A) == getID(C) || getID(B) == getID(C) ) {
+#ifdef _DEBUG
+				std::cerr << "grb::eWiseApply: The output matrix can not simultaneously be "
+					<< "one of the input matrices\n";
+#endif
+			}
+
+			// run-time checks
+			const size_t m = nrows( C );
+			const size_t n = ncols( C );
+			const size_t m_A = !trans_left ? nrows( A ) : ncols( A );
+			const size_t n_A = !trans_left ? ncols( A ) : nrows( A );
+			const size_t m_B = !trans_right ? nrows( B ) : ncols( B );
+			const size_t n_B = !trans_right ? ncols( B ) : nrows( B );
+
+			// Identities
+			const auto identity_A = monoid.template getIdentity< OutputType >();
+			const auto identity_B = monoid.template getIdentity< OutputType >();
+
+			if( m != m_A || m != m_B || n != n_A || n != n_B ) {
 #ifdef _DEBUG
-						std::cout << "A( " << i << ", " << k_col << " ) = " << A_raw.getValue( k,
-							mulMonoid.template getIdentity< typename Operator::D1 >() ) << ", ";
+				std::cerr << "grb::eWiseApply: dimensions of input matrices do not match\n";
 #endif
+				return MISMATCH;
+			}
+
+			const auto oper = monoid.getOperator();
+			const auto &A_raw = !trans_left ?
+				internal::getCRS( A ) :
+				internal::getCCS( A );
+			const auto &B_raw = !trans_right ?
+				internal::getCRS( B ) :
+				internal::getCCS( B );
+			auto &CRS_raw = internal::getCRS( C );
+			auto &CCS_raw = internal::getCCS( C );
+
+
+			// retrieve buffers
+			char *arr1 = nullptr, *arr3 = nullptr;
+			char *buf1 = nullptr, *buf3 = nullptr;
+			InputType1 * vbuf1 = nullptr;
+			OutputType * vbuf3 = nullptr;
+			internal::getMatrixBuffers( arr1, buf1, vbuf1, 1, A );
+			internal::getMatrixBuffers( arr3, buf3, vbuf3, 1, C );
+			// end buffer retrieval
+
+			// initialisations of the coordinates
+			// Note: By using the buffer of the output matrix C, we can
+			//       allow A and B to be the same matrix (with the same buffer)
+			Coordinates< reference > coors1, coors2;
+			coors1.set( arr1, false, buf1, n );
+			coors2.set( arr3, false, buf3, n );
+			// end initialisations of the coordinates
+
+			if( !crs_only ) {
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+				#pragma omp parallel for simd default(none) shared(CCS_raw) firstprivate(n)
+#endif
+				for( size_t j = 0; j < n + 1; ++j ) {
+					CCS_raw.col_start[ j ] = 0;
+				}
+			}
+			// end initialisations
+
+			// nonzero count
+			size_t nzc = 0;
+
+			// symbolic phase
+			if( phase == RESIZE ) {
+				nzc = 0;
+
+
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+				#pragma omp parallel default(none) \
+					shared(coors1, vbuf1, coors2, vbuf3) \
+					firstprivate(A_raw, identity_A, B_raw, identity_B, m) \
+					reduction(+:nzc)
+#endif
+				{
+					for( size_t i = 0; i < m; ++i ) {
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+						#pragma omp single
+#endif
+						{
+							coors1.clear();
+						}
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+						#pragma omp barrier
```

Comment:
I would expect it to be cheaper to start the parallel section at line 1378, rather than using a single for the clear (also does the single not imply a barrier at the end of it?)